### PR TITLE
ARROW-11212: [Packaging][Python] Use vcpkg as dependency source for manylinux and windows wheels

### PR DESCRIPTION
**Pull request by kszucs**
Also resolves:
- ARROW-11213: [Packaging][Python] Dockerize wheel building on windows
- ARROW-11215: [CI] Use named volumes by default for caching in docker-compose
- ARROW-11231: [Python][Packaging] Enable mimalloc in wheels 

Main features:
- dockerize windows builds
- use vcpkg as the dependency source where we can explicitly pin a working version without worrying about the drag of build environment over time

Potential follow-up:
- use vcpkg on macos as well

cc @kou @xhochy 

## Manylinux testing

Should be straightforward (I'm going to shared pre-built images for quicker testing).

## Windows testing

Only windows host is able to run windows containers. I'm virtualizing windows on macOS (should work on linux as well) using virtualbox:

```bash
git clone https://github.com/StefanScherer/windows-docker-machine
cd windows-docker-machine
# grant more resources in the vagrantfile under the virtualbox section (I use 8 cores and 16GiB of ram)
vagrant up --provider virtualbox 2019-box
docker context ls
docker context use 2019-box
docker image ls
```

Now docker should use the windows docker daemon. 

```bash
pip install -e dev/archery[docker]
# the volumes defined in the docker compose file should be commented out and pass arrow's source explicitly
# this is required because compose is being executed on a unix host and the directory is mounted through a virtual machine
PYTHON=3.6 archery docker run --no-pull --using-docker-cli -v C:$(pwd):C:/arrow python-wheel-windows-vs2017 cmd.exe
# then within the container execute a build or do other interactive things
arrow\ci\scripts\python_wheel_windown_build.bat
```

### CODE REVIEW

**Comment by kou** [comment]
Related works:

  * https://github.com/apache/arrow/pull/8386#issuecomment-728321361
  * #8881

**Comment by kszucs** [comment]
Currently integrating with crossbow using pre-built images, the manylinux build already works except the artifact uploading.
Pull & build took ~13m: https://github.com/ursa-labs/crossbow/runs/1652174767

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-manylinux*

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-manylinux*

**Comment by kszucs** [comment]
@xhochy we have a [running build](https://github.com/ursa-labs/crossbow/runs/1658737075) for windows on github actions now. I'll do a couple of things to try to improve the build times a bit (make the image smaller because pulling takes ~15m and enable unity builds). 

Since we don't bundle `msvcp140.dll` these wheels are not working in the official windows python docker containers, so I'm using the same image as for building. I may try to bundle it (guess it should be about copying it into the archive) to ensure that it works in the base images.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-manylinux2014-cp36m wheel-windows-cp36m

**Comment by kszucs** [comment]
@kou @xhochy bundled the msvc runtime. Locally tested with the stock windows python images and seems to work fine. Hopefully crossbow will confirm it.

I'm going to write a more detailed description of the changes tomorrow. Any reviews are welcome!

**Comment by kou** [comment]
It seems that https://github.com/ursa-labs/crossbow/releases/tag/actions-838-github-wheel-windows-cp36m doesn't have artifacts...
Where did you download built wheels?

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-windows-cp36m

**Comment by kszucs** [comment]
The build also executes the tests in a different docker container, see the "Test Wheel" step in the [build](https://github.com/ursa-labs/crossbow/runs/1664830035?check_suite_focus=true)

File: ci/scripts/wheel_windows_build.bat
Comment (kou) [reply ci/scripts/wheel_windows_build.bat:90]: Ah, Python for Windows includes `vcruntime140.dll`.
We don't need to bundle them.

**Review by kou — APPROVED** [review]
+1

File: ci/scripts/wheel_windows_build.bat
Comment (kszucs) [reply ci/scripts/wheel_windows_build.bat:90]: I plan to implement this bundling logic in cmake, but we can defer it for after the release.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-manylinux2014-cp36m wheel-windows-cp36m

File: python/setup.py
Comment (kszucs) [inline python/setup.py:388]: We can choose to bundle security sensitive shared libraries instead of linking them statically.

File: dev/tasks/tasks.yml
Comment (kszucs) [inline dev/tasks/tasks.yml:517]: `choco` fails to install python 3.9, investigating it.

File: python/setup.py
Comment (xhochy) [reply python/setup.py:388]: In the case of OpenSSL, if we bundled a shared version the version in the main system will be preferred. Not sure though where this is updated from. In `conda-forge` we explicitly have a Python setup that takes care that one is not used.

Comment (kszucs) [reply python/setup.py:388]: > In the case of OpenSSL, if we bundled a shared version the version in the main system will be preferred.

Shouldn't we prefer the system one over the bundled one?

> Not sure though where this is updated from.

You mean the bundled or the system one?

Comment (xhochy) [reply python/setup.py:388]: Not sure about this. I don't know where the system one is coming from.

Comment (kszucs) [reply python/setup.py:388]: How about keeping it as is, linking these libraries statically?

Comment (xhochy) [reply python/setup.py:388]: We should keep as-is until someone visits with a strong *and*  reasonable opinion.

File: dev/tasks/tasks.yml
Comment (kszucs) [reply dev/tasks/tasks.yml]: Fixed.

File: python/setup.py
Comment (kszucs) [reply python/setup.py:388]: Ok, I'm going to clean up the commented section used for bundling from `python/CMakeLists.txt` and `setup.py` then.

Comment (kszucs) [reply python/setup.py:388]: And create a follow-up JIRA.

File: ci/scripts/wheel_windows_build.bat
Comment (kszucs) [reply ci/scripts/wheel_windows_build.bat:90]: Created a follow-up jira https://issues.apache.org/jira/browse/ARROW-11185

File: python/setup.py
Comment (kszucs) [reply python/setup.py:388]: https://issues.apache.org/jira/browse/ARROW-11186

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-*

**Comment by kszucs** [comment]
The osx builds shouldn't be failing, but that's out of the scope of this PR.

**Comment by xhochy** [comment]
The OSX issue is that `protobuf` is installed through `brew` and that doesn't install it using `cmake`.  You could conditionally add the following line https://github.com/apache/arrow/blob/a576cd71c08c3fec029d24092c407ece1775f5a6/cpp/cmake_modules/ThirdpartyToolchain.cmake#L2459 only when protobuf is also build using bundled.

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:139]: How about prepend `python-` because `python-` exists in related names?

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:160]: How about prepend `python-` because `python-` exists in related names?

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:169]: How about prepend `python_` because `python-` exists in related names?

File: docker-compose.yml
Comment (kou) [inline docker-compose.yml:225]: How about moving `python-wheel-*` services into `...### Python ###...` section?

File: docker-compose.yml
Comment (kszucs) [reply docker-compose.yml]: Updated.

File: docker-compose.yml
Comment (kszucs) [reply docker-compose.yml]: Updated.

File: docker-compose.yml
Comment (kszucs) [reply docker-compose.yml]: Updated.

File: docker-compose.yml
Comment (kszucs) [reply docker-compose.yml]: Updated.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-osx-mavericks-cp36m wheel-osx-high-sierra-cp36m

File: .env
Comment (kszucs) [inline .env:25]: Need to set this in the github actions workflow where docker-compose is used.

**Comment by kszucs** [comment]
@github-actions crossbow -c ursacomputing/crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow -c ursacomputing/crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow -c ursacomputing/crossbow submit -g wheel

File: python/pyarrow/tests/test_fs.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_fs.py:1332]: Should this be reverted before merging? 
(I thought @xhochy encountered a similar some time ago?)

Comment (kszucs) [reply python/pyarrow/tests/test_fs.py:1332]: Nice catch, definitely.

**Review by xhochy — APPROVED** [review]
LGTM from my side

File: ci/scripts/python_wheel_manylinux_build.sh
Comment (pitrou) [inline ci/scripts/python_wheel_manylinux_build.sh:105]: Switch to ON to make builds slightly faster.

Comment (kszucs) [reply ci/scripts/python_wheel_manylinux_build.sh]: Updated.

**Comment by kszucs** [comment]
I'm building the images and submitting a crossbow job afterwards.

**Comment by pitrou** [comment]
@kszucs We would need to apply the following patch when building snappy in order to solve ARROW-7939. Is it possible?
https://github.com/conda-forge/snappy-feedstock/blob/5b7b850667c769df4551c81dc8ffaef76e8d371b/recipe/disable_bmi.patch

**Comment by kszucs** [comment]
@github-actions crossbow -c ursacomputing/crossbow submit -g wheel

File: dev/tasks/python-wheels/osx-build.sh
Comment (kou) [inline dev/tasks/python-wheels/osx-build.sh:95]: I think that this is needless.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2473]: How about the following?

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index e5c72be89..1ebf836a3 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1326,6 +1326,7 @@ endif()
 
 macro(build_protobuf)
   message("Building Protocol Buffers from source")
+  set(PROTOBUF_VENDORED TRUE)
   set(PROTOBUF_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/protobuf_ep-install")
   set(PROTOBUF_INCLUDE_DIR "${PROTOBUF_PREFIX}/include")
   # Newer protobuf releases always have a lib prefix independent from CMAKE_STATIC_LIBRARY_PREFIX
@@ -2466,13 +2467,15 @@ macro(build_grpc)
       -DgRPC_GFLAGS_PROVIDER=package
       -DgRPC_MSVC_STATIC_RUNTIME=${ARROW_USE_STATIC_CRT}
       -DgRPC_PROTOBUF_PROVIDER=package
-      -DgRPC_PROTOBUF_PACKAGE_TYPE=CONFIG
       -DgRPC_RE2_PROVIDER=package
       -DgRPC_SSL_PROVIDER=package
       -DgRPC_ZLIB_PROVIDER=package
       -DCMAKE_INSTALL_PREFIX=${GRPC_PREFIX}
       -DCMAKE_INSTALL_LIBDIR=lib
       -DBUILD_SHARED_LIBS=OFF)
+  if(PROTOBUF_VENDORED)
+    list(APPEND GRPC_CMAKE_ARGS -DgRPC_PROTOBUF_PACKAGE_TYPE=CONFIG)
+  endif()
   if(OPENSSL_ROOT_DIR)
     list(APPEND GRPC_CMAKE_ARGS -DOPENSSL_ROOT_DIR=${OPENSSL_ROOT_DIR})
   endif()
```

**Comment by pitrou** [comment]
Hmm, the SSL certificate check failure on manylinux seems serious and should be fixed. It may be due to this line:
https://github.com/microsoft/vcpkg/blob/master/ports/openssl/unix/CMakeLists.txt#L171

**Comment by pitrou** [comment]
Also, looking at https://github.com/microsoft/vcpkg/blob/master/ports/openssl/unix/CMakeLists.txt#L8-L13, it does not seem Linux builds would be supported on non-x86 platforms (such as arm64).

**Comment by kszucs** [comment]
@pitrou you have resolved it previously with https://github.com/pitrou/arrow/commit/b0dd66f17e94fbf61da0f1f616ccd5e627a79145

**Comment by kszucs** [comment]
@github-actions crossbow -c ursacomputing/crossbow submit -g packaging

**Comment by kszucs** [comment]
I submitted the crossbow tasks using the CLI, the in-progress results are promising:
https://github.com/ursacomputing/crossbow/branches/all?query=build-17

**Comment by kszucs** [comment]
@kou this should be ready to merge. One of the arm linux builds is [failing with a timeout error](https://github.com/ursacomputing/crossbow/runs/1704000375) in addition to the arm conda drone builds, but these shouldn't block this PR. The latest crossbow submission is here https://github.com/ursacomputing/crossbow/branches/all?query=build-18

**Comment by kou** [comment]
Thanks!

I'll merge this. We need to work on some follow-up tasks:

* Fix .deb/.rpm builds on Travis CI
* Fix Python version (3.8) in `dev/tasks/conda-recipes/.ci_support/osx_pytho3.7.____cpython.yaml`

**Pull request closed without merging**
